src/java/org/apache/cassandra/db/RangeTombstoneList.java

```diff
@@ -181,23 +192,44 @@ public void addAll(RangeTombstoneList tombstones)
         }
         else
         {
-            int i = 0;
-            int j = 0;
+            int i = 0; // pointer into "this"
+            int j = 0; // pointer into "tombstones"
+            
+            // Verify that this is sorted:
+            tombstones.assertSorted();
+            assertSorted();
+            
+            // basically this is a linear search for the position in "this" where tombstones.starts[j] should be inserted.
+            // It's written such that it only scans each list once, since they're both supposed to be sorted.
             while (i < size && j < tombstones.size)
             {
                 if (comparator.compare(tombstones.starts[j], ends[i]) <= 0)
-                {
+                {   
+                    // ``tombstones`` item is before ends[i], so insert it there.
                     insertFrom(i, tombstones.starts[j], tombstones.ends[j], tombstones.markedAts[j], tombstones.delTimes[j]);
-                    j++;
+                    j++; // and move to the next item to insert.
+                    assertSorted();
                 }
                 else
                 {
-                    i++;
+                    i++; // ``tombstones`` item is after ends[i], so 
                 }
             }
-            // Addds the remaining ones from tombstones if any (note that addInternal will increment size if relevant).
+            // Adds the remaining items from ``tombstones`` if any (note that addInternal will increment size if relevant).
             for (; j < tombstones.size; j++)
                 addInternal(size, tombstones.starts[j], tombstones.ends[j], tombstones.markedAts[j], tombstones.delTimes[j]);
+            assertSorted();
+        }
+    }
+
+    /**
+     * Check the invariant for RangeTombstoneList: that all items are in order.
+     */
+    private void assertSorted() {
+        for(int p = 0; p < size -1; ++p) {
```

Comment:
They could be as large as the number of "active" tombstones. So on the order of thousands. We wouldn't want to run this in prod *unless* we were hunting existing corruption.